roachtest: rebalance/by-load/replicas/mixed-version failed

**Note:** This build has runtime assertions enabled. If the same failure was hit in a run without assertions enabled, there should be a similar failure without this message. If there isn&#39;t one, then this failure is likely due to an assertion violation or (assertion) timeout.

roachtest.rebalance/by-load/replicas/mixed-version [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyAzureBazel/19636120?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyAzureBazel/19636120?buildTab=artifacts#/rebalance/by-load/replicas/mixed-version) on master @ [ea65dae5f264c2e9ffb5846182e0740d487ca0b3](https://github.com/cockroachdb/cockroach/commits/ea65dae5f264c2e9ffb5846182e0740d487ca0b3):


```
(mixedversion.go:829).Run: unexpected node event: n6: cockroach process for system interface died (exit code 134)
test artifacts and logs in: /artifacts/rebalance/by-load/replicas/mixed-version/run_1
```

Parameters:
 - <code>arch=amd64</code>
 - <code>cloud=azure</code>
 - <code>coverageBuild=false</code>
 - <code>cpu=4</code>
 - <code>encrypted=false</code>
 - <code>fs=ext4</code>
 - <code>localSSD=true</code>
 - <code>metamorphicBufferedSender=false</code>
 - <code>mvtDeploymentMode=shared-process</code>
 - <code>mvtVersions=v24.2.10 → v24.3.12 → master</code>
 - <code>runtimeAssertionsBuild=true</code>
 - <code>ssd=0</code>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

_Grafana is not yet available for azure clusters_</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*rebalance/by-load/replicas/mixed-version.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-50458

Crash on 24.3.12:

```
I250507 08:59:19.862661 1 util/log/file_sync_buffer.go:237 â‹® [config]   binary: CockroachDB CCL v24.3.12 (x86_64-pc-linux-gnu, built 2025/04/28 13:25:26, go1.22.8 X:nocoverageredesign)
```

```
I250507 08:59:19.862452 1 util/log/flags.go:225  [-] 1  stderr capture started
panic: runtime error: index out of range [8192] with length 8192 [recovered]
	panic: runtime error: index out of range [8192] with length 8192 [recovered]
	panic: runtime error: index out of range [8192] with length 8192

goroutine 51504 gp=0xc002f2c1c0 m=11 mp=0xc013530808 [running]:
panic({0x69bf0a0?, 0xc01927e0a8?})
	GOROOT/src/runtime/panic.go:779 +0x158 fp=0xc005908278 sp=0xc0059081c8 pc=0x49c778
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0x80?, {0x8698000, 0xc01c096510})
	pkg/util/stop/stopper.go:226 +0x65 fp=0xc0059082c0 sp=0xc005908278 pc=0x131e025
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2.deferwrap4()
	pkg/util/stop/stopper.go:492 +0x28 fp=0xc0059082e8 sp=0xc0059082c0 pc=0x131f9e8
panic({0x69bf0a0?, 0xc01927e0a8?})
	GOROOT/src/runtime/panic.go:770 +0x132 fp=0xc005908398 sp=0xc0059082e8 pc=0x49c752
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Wait(0xc01cd57188)
	pkg/sql/flowinfra/flow.go:609 +0x19b fp=0xc005908428 sp=0xc005908398 pc=0x2a9d75b
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run.deferwrap1()
	pkg/sql/flowinfra/flow.go:551 +0x25 fp=0xc005908440 sp=0xc005908428 pc=0x2a9d585
panic({0x69bf0a0?, 0xc01927e0a8?})
	GOROOT/src/runtime/panic.go:770 +0x132 fp=0xc0059084f0 sp=0xc005908440 pc=0x49c752
runtime.goPanicIndex(0x2000, 0x2000)
	GOROOT/src/runtime/panic.go:114 +0x7c fp=0xc005908530 sp=0xc0059084f0 pc=0x49b17c
github.com/axiomhq/hyperloglog.(*registers).get(...)
	external/com_github_axiomhq_hyperloglog/registers.go:80
github.com/axiomhq/hyperloglog.(*Sketch).Merge(0xc01cf78780, 0xc005908668)
	external/com_github_axiomhq_hyperloglog/hyperloglog.go:157 +0x3a7 fp=0xc005908620 sp=0xc005908530 pc=0x2946147
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*sampleAggregator).processSketchRow(0xc01cf86008, 0xc01cf66120, {0xc006b250e8, 0xa, 0xc006b24f08?}, 0xc005908768)
	pkg/sql/rowexec/sample_aggregator.go:398 +0x24a fp=0xc0059086a8 sp=0xc005908620 pc=0x297c96a
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*sampleAggregator).mainLoop(0xc01cf86008, {0x8698000, 0xc01cf788d0}, {0x865d610, 0xc00ff97c08})
	pkg/sql/rowexec/sample_aggregator.go:350 +0x56a fp=0xc005908ae8 sp=0xc0059086a8 pc=0x297c08a
```

It's unclear to me whether we're "just accepting these crashes" for some reason or whether this means that  https://github.com/cockroachdb/cockroach/issues/133054#issuecomment-2580830693 was insufficient.

Reassigning to SQL Queries.  (cc @yuzefovich)

--- a/DEPS.bzl
+++ b/DEPS.bzl
@@ -784,6 +784,16 @@ def go_deps():
         ],
     )
     go_repository(
+        name = "com_github_axiomhq_hyperloglog_000",
+        build_file_proto_mode = "disable_global",
+        importpath = "github.com/axiomhq/hyperloglog/000",
+        sha256 = "812834322ee2ca50dc36f91f9ac3f2cde4631af2f9c330b1271c78b46024a540",
+        strip_prefix = "github.com/axiomhq/hyperloglog@v0.0.0-20181223111420-4b99d0c2c99e",
+        urls = [
+            "https://storage.googleapis.com/cockroach-godeps/gomod/github.com/axiomhq/hyperloglog/com_github_axiomhq_hyperloglog-v0.0.0-20181223111420-4b99d0c2c99e.zip",
+        ],
+    )
+    go_repository(
         name = "com_github_aymanbagabas_go_osc52",
         build_file_proto_mode = "disable_global",
         importpath = "github.com/aymanbagabas/go-osc52",

--- a/build/bazelutil/distdir_files.bzl
+++ b/build/bazelutil/distdir_files.bzl
@@ -289,6 +289,7 @@ DISTDIR_FILES = {
     "https://storage.googleapis.com/cockroach-godeps/gomod/github.com/aws/aws-sdk-go-v2/service/sts/com_github_aws_aws_sdk_go_v2_service_sts-v1.33.17.zip": "87aca25fafd483a1eac29c5baaab05ad485422a9aa1ccc5db0d39733c2d71cd2",
     "https://storage.googleapis.com/cockroach-godeps/gomod/github.com/aws/aws-sdk-go/com_github_aws_aws_sdk_go-v1.40.37.zip": "c0c481d28af88f621fb3fdeacc1e5d32f69a1bb83d0ee959f95ce89e4e2d0494",
     "https://storage.googleapis.com/cockroach-godeps/gomod/github.com/aws/smithy-go/com_github_aws_smithy_go-v1.22.3.zip": "572df48de9133d57f45909d3067b2053b97230268c2d28e4e44ea9644009ef11",
+    "https://storage.googleapis.com/cockroach-godeps/gomod/github.com/axiomhq/hyperloglog/com_github_axiomhq_hyperloglog-v0.0.0-20181223111420-4b99d0c2c99e.zip": "812834322ee2ca50dc36f91f9ac3f2cde4631af2f9c330b1271c78b46024a540",
     "https://storage.googleapis.com/cockroach-godeps/gomod/github.com/axiomhq/hyperloglog/com_github_axiomhq_hyperloglog-v0.2.5.zip": "6125b12664bb5dd8614e82f0fe7528242dcb11649e1d7e051aabf3da471e14e1",
     "https://storage.googleapis.com/cockroach-godeps/gomod/github.com/aymanbagabas/go-osc52/com_github_aymanbagabas_go_osc52-v1.0.3.zip": "138e75a51599c2a8e4afe2bd6acdeaddbb73eb9ec796dfa2f577b16201660d9e",
     "https://storage.googleapis.com/cockroach-godeps/gomod/github.com/aymerick/douceur/com_github_aymerick_douceur-v0.2.0.zip": "dcbf69760cc1a8b32384495438e1086e4c3d669b2ebc0debd92e1865ffd6be60",

--- a/go.mod
+++ b/go.mod
@@ -123,6 +123,7 @@ require (
 	github.com/aws/aws-sdk-go-v2/service/secretsmanager v1.35.1
 	github.com/aws/smithy-go v1.22.3
 	github.com/axiomhq/hyperloglog v0.2.5
+	github.com/axiomhq/hyperloglog/000 v0.0.0-20181223111420-4b99d0c2c99e
 	github.com/bazelbuild/rules_go v0.26.0
 	github.com/biogo/store v0.0.0-20160505134755-913427a1d5e8
 	github.com/blevesearch/snowballstem v0.9.0
@@ -507,6 +508,10 @@ replace golang.org/x/time => github.com/cockroachdb/x-time v0.3.1-0.202305251236
 replace github.com/gogo/protobuf => github.com/cockroachdb/gogoproto v1.3.3-0.20241216150617-2358cdb156a1
 
 replace storj.io/drpc => github.com/cockroachdb/drpc v0.0.0-20240709171858-0075ac871661
+
+// TODO(yuzefovich): remove this version once compatibility with 24.3 is no
+// longer needed.
+replace github.com/axiomhq/hyperloglog/000 => github.com/axiomhq/hyperloglog v0.0.0-20181223111420-4b99d0c2c99e
 
 // Note: This forked dependency adds a commit that opens up some
 // private APIs to enable us to make some perf improvements to

--- a/pkg/sql/distsql_plan_stats.go
+++ b/pkg/sql/distsql_plan_stats.go
@@ -410,6 +410,7 @@ func (dsp *DistSQLPlanner) createPartialStatsPlan(
 
 	sampledColumnIDs := make([]descpb.ColumnID, len(scan.catalogCols))
 	spec := execinfrapb.SketchSpec{
+		SketchType:          execinfrapb.SketchType_HLL_PLUS_PLUS_V1,
 		GenerateHistogram:   reqStat.histogram,
 		HistogramMaxBuckets: reqStat.histogramMaxBuckets,
 		Columns:             make([]uint32, len(reqStat.columns)),
@@ -655,6 +656,7 @@ func (dsp *DistSQLPlanner) createStatsPlan(
 	sampledColumnIDs := make([]descpb.ColumnID, len(requestedCols))
 	for _, s := range reqStats {
 		spec := execinfrapb.SketchSpec{
+			SketchType:          execinfrapb.SketchType_HLL_PLUS_PLUS_V1,
 			GenerateHistogram:   s.histogram,
 			HistogramMaxBuckets: s.histogramMaxBuckets,
 			Columns:             make([]uint32, len(s.columns)),

--- a/pkg/sql/execinfrapb/processors_table_stats.proto
+++ b/pkg/sql/execinfrapb/processors_table_stats.proto
@@ -17,9 +17,18 @@ option go_package = "github.com/cockroachdb/cockroach/pkg/sql/execinfrapb";
 import "sql/catalog/descpb/structured.proto";
 import "gogoproto/gogo.proto";
 
+// TODO(yuzefovich): this can be removed once compatibility with 24.3 is no
+// longer needed.
+enum SketchType {
+  // This is the github.com/axiomhq/hyperloglog binary format (as of commit
+  // 730eea1) for a sketch with precision 14. Values are encoded using their key
+  // encoding, except integers which are encoded in 8 bytes (little-endian).
+  HLL_PLUS_PLUS_V1 = 0;
+}
+
 // SketchSpec contains the specification for a generated statistic.
 message SketchSpec {
-  reserved 1;
+  optional SketchType sketch_type = 1 [(gogoproto.nullable) = false];
 
   // Each value is an index identifying a column in the input stream.
   // TODO(radu): currently only one column is supported.

--- a/pkg/sql/logictest/logictestbase/logictestbase.go
+++ b/pkg/sql/logictest/logictestbase/logictestbase.go
@@ -649,8 +649,6 @@ var DefaultConfigSets = map[string]ConfigSet{
 
 	// Special alias for all testserver configs (for mixed-version testing).
 	"cockroach-go-testserver-configs": makeConfigSet(
-		// TODO(yuzefovich): when removing 24.3 version from here, ensure to
-		// use the alias in mixed_version_stats.
 		"cockroach-go-testserver-24.3",
 		"cockroach-go-testserver-25.1",
 	),

--- a/pkg/sql/logictest/testdata/logic_test/mixed_version_stats
+++ b/pkg/sql/logictest/testdata/logic_test/mixed_version_stats
@@ -1,4 +1,4 @@
-# LogicTest: cockroach-go-testserver-25.1
+# LogicTest: cockroach-go-testserver-configs
 
 # This is a regression test for #137386 which verifies the hyperloglog library
 # upgrade for the table stats. We want such a distributed table stats collection

--- a/pkg/sql/logictest/tests/cockroach-go-testserver-24.3/BUILD.bazel
+++ b/pkg/sql/logictest/tests/cockroach-go-testserver-24.3/BUILD.bazel
@@ -11,7 +11,7 @@ go_test(
         "//pkg/sql/logictest:testdata",  # keep
     ],
     exec_properties = {"test.Pool": "heavy"},
-    shard_count = 10,
+    shard_count = 11,
     tags = ["cpu:3"],
     deps = [
         "//pkg/base",

--- a/pkg/sql/logictest/tests/cockroach-go-testserver-24.3/generated_test.go
+++ b/pkg/sql/logictest/tests/cockroach-go-testserver-24.3/generated_test.go
@@ -115,6 +115,13 @@ func TestLogic_mixed_version_plpgsql_srf(
 	runLogicTest(t, "mixed_version_plpgsql_srf")
 }
 
+func TestLogic_mixed_version_stats(
+	t *testing.T,
+) {
+	defer leaktest.AfterTest(t)()
+	runLogicTest(t, "mixed_version_stats")
+}
+
 func TestLogic_mixed_version_timeseries_range_already_exists(
 	t *testing.T,
 ) {

--- a/pkg/sql/rowexec/BUILD.bazel
+++ b/pkg/sql/rowexec/BUILD.bazel
@@ -115,6 +115,7 @@ go_library(
         "//pkg/util/tracing/tracingpb",
         "//pkg/util/vector",
         "@com_github_axiomhq_hyperloglog//:hyperloglog",
+        "@com_github_axiomhq_hyperloglog_000//:000",
         "@com_github_cockroachdb_errors//:errors",
         "@com_github_cockroachdb_logtags//:logtags",
         "@com_github_cockroachdb_redact//:redact",

--- a/pkg/sql/rowexec/sample_aggregator.go
+++ b/pkg/sql/rowexec/sample_aggregator.go
@@ -10,12 +10,14 @@ import (
 	"math"
 	"time"
 
-	"github.com/axiomhq/hyperloglog"
+	hllNew "github.com/axiomhq/hyperloglog"
+	hllOld "github.com/axiomhq/hyperloglog/000"
 	"github.com/cockroachdb/cockroach/pkg/jobs"
 	"github.com/cockroachdb/cockroach/pkg/server/telemetry"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
 	"github.com/cockroachdb/cockroach/pkg/sql/execinfra"
 	"github.com/cockroachdb/cockroach/pkg/sql/execinfrapb"
+	"github.com/cockroachdb/cockroach/pkg/sql/execversion"
 	"github.com/cockroachdb/cockroach/pkg/sql/isql"
 	"github.com/cockroachdb/cockroach/pkg/sql/parser"
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode"
@@ -101,6 +103,7 @@ func newSampleAggregator(
 			return nil, errors.Errorf("histograms require one column")
 		}
 	}
+	useNewHLL := execversion.FromContext(ctx) >= execversion.V25_1
 
 	// Limit the memory use by creating a child monitor with a hard limit.
 	// The processor will disable histogram collection if this limit is not
@@ -141,9 +144,13 @@ func newSampleAggregator(
 	for i := range spec.Sketches {
 		s.sketches[i] = sketchInfo{
 			spec:     spec.Sketches[i],
-			sketch:   hyperloglog.New14(),
 			numNulls: 0,
 			numRows:  0,
+		}
+		if useNewHLL {
+			s.sketches[i].sketchNew = hllNew.New14()
+		} else {
+			s.sketches[i].sketchOld = hllOld.New14()
 		}
 		if spec.Sketches[i].GenerateHistogram {
 			sampleCols.Add(int(spec.Sketches[i].Columns[0]))
@@ -166,9 +173,13 @@ func newSampleAggregator(
 		s.invSr[col] = &sr
 		s.invSketch[col] = &sketchInfo{
 			spec:     spec.InvertedSketches[i],
-			sketch:   hyperloglog.New14(),
 			numNulls: 0,
 			numRows:  0,
+		}
+		if useNewHLL {
+			s.invSketch[col].sketchNew = hllNew.New14()
+		} else {
+			s.invSketch[col].sketchOld = hllOld.New14()
 		}
 	}
 
@@ -361,8 +372,6 @@ func (s *sampleAggregator) mainLoop(
 func (s *sampleAggregator) processSketchRow(
 	sketch *sketchInfo, row rowenc.EncDatumRow, da *tree.DatumAlloc,
 ) error {
-	var tmpSketch hyperloglog.Sketch
-
 	numRows, err := row[s.numRowsCol].GetInt()
 	if err != nil {
 		return err
@@ -389,11 +398,22 @@ func (s *sampleAggregator) processSketchRow(
 	if d == tree.DNull {
 		return errors.AssertionFailedf("NULL sketch data")
 	}
-	if err := tmpSketch.UnmarshalBinary([]byte(*d.(*tree.DBytes))); err != nil {
-		return err
-	}
-	if err := sketch.sketch.Merge(&tmpSketch); err != nil {
-		return errors.NewAssertionErrorWithWrappedErrf(err, "merging sketch data")
+	if sketch.sketchNew != nil {
+		var tmpSketch hllNew.Sketch
+		if err := tmpSketch.UnmarshalBinary([]byte(*d.(*tree.DBytes))); err != nil {
+			return err
+		}
+		if err := sketch.sketchNew.Merge(&tmpSketch); err != nil {
+			return errors.NewAssertionErrorWithWrappedErrf(err, "merging sketch data")
+		}
+	} else {
+		var tmpSketch hllOld.Sketch
+		if err := tmpSketch.UnmarshalBinary([]byte(*d.(*tree.DBytes))); err != nil {
+			return err
+		}
+		if err := sketch.sketchOld.Merge(&tmpSketch); err != nil {
+			return errors.NewAssertionErrorWithWrappedErrf(err, "merging sketch data")
+		}
 	}
 	return nil
 }
@@ -611,7 +631,12 @@ func (s *sampleAggregator) getAvgSize(si *sketchInfo) int64 {
 // getDistinctCount returns the number of distinct values in the given sketch,
 // optionally including null values.
 func (s *sampleAggregator) getDistinctCount(si *sketchInfo, includeNulls bool) int64 {
-	distinctCount := int64(si.sketch.Estimate())
+	var distinctCount int64
+	if si.sketchNew != nil {
+		distinctCount = int64(si.sketchNew.Estimate())
+	} else {
+		distinctCount = int64(si.sketchOld.Estimate())
+	}
 	if si.numNulls > 0 && !includeNulls {
 		// Nulls are included in the estimate, so reduce the count by 1 if nulls are
 		// not requested.

--- a/pkg/sql/rowexec/sampler.go
+++ b/pkg/sql/rowexec/sampler.go
@@ -11,7 +11,8 @@ import (
 	"math/rand"
 	"time"
 
-	"github.com/axiomhq/hyperloglog"
+	hllNew "github.com/axiomhq/hyperloglog"
+	hllOld "github.com/axiomhq/hyperloglog/000"
 	"github.com/cockroachdb/cockroach/pkg/server/telemetry"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/catenumpb"
 	"github.com/cockroachdb/cockroach/pkg/sql/execinfra"
@@ -35,8 +36,10 @@ import (
 
 // sketchInfo contains the specification and run-time state for each sketch.
 type sketchInfo struct {
-	spec                 execinfrapb.SketchSpec
-	sketch               *hyperloglog.Sketch
+	spec execinfrapb.SketchSpec
+	// Exactly one of sketchOld and sketchNew will be set.
+	sketchOld            *hllOld.Sketch
+	sketchNew            *hllNew.Sketch
 	numNulls             int64
 	numRows              int64
 	size                 int64
@@ -100,6 +103,7 @@ func newSamplerProcessor(
 	input execinfra.RowSource,
 	post *execinfrapb.PostProcessSpec,
 ) (*samplerProcessor, error) {
+	useNewHLL := execversion.FromContext(ctx) >= execversion.V25_1
 	legacyFingerprinting := execversion.FromContext(ctx) < execversion.V25_2
 
 	// Limit the memory use by creating a child monitor with a hard limit.
@@ -129,10 +133,14 @@ func newSamplerProcessor(
 	for i := range spec.Sketches {
 		s.sketches[i] = sketchInfo{
 			spec:                 spec.Sketches[i],
-			sketch:               hyperloglog.New14(),
 			numNulls:             0,
 			numRows:              0,
 			legacyFingerprinting: legacyFingerprinting,
+		}
+		if useNewHLL {
+			s.sketches[i].sketchNew = hllNew.New14()
+		} else {
+			s.sketches[i].sketchOld = hllOld.New14()
 		}
 		if spec.Sketches[i].GenerateHistogram {
 			sampleCols.Add(int(spec.Sketches[i].Columns[0]))
@@ -152,9 +160,13 @@ func newSamplerProcessor(
 		sketchSpec.Columns = []uint32{0}
 		s.invSketch[col] = &sketchInfo{
 			spec:     sketchSpec,
-			sketch:   hyperloglog.New14(),
 			numNulls: 0,
 			numRows:  0,
+		}
+		if useNewHLL {
+			s.invSketch[col].sketchNew = hllNew.New14()
+		} else {
+			s.invSketch[col].sketchOld = hllOld.New14()
 		}
 	}
 
@@ -425,7 +437,12 @@ func (s *samplerProcessor) emitSketchRow(
 	outRow[s.numRowsCol] = rowenc.EncDatum{Datum: tree.NewDInt(tree.DInt(si.numRows))}
 	outRow[s.numNullsCol] = rowenc.EncDatum{Datum: tree.NewDInt(tree.DInt(si.numNulls))}
 	outRow[s.sizeCol] = rowenc.EncDatum{Datum: tree.NewDInt(tree.DInt(si.size))}
-	data, err := si.sketch.MarshalBinary()
+	var data []byte
+	if si.sketchNew != nil {
+		data, err = si.sketchNew.MarshalBinary()
+	} else {
+		data, err = si.sketchOld.MarshalBinary()
+	}
 	if err != nil {
 		return false, err
 	}
@@ -586,7 +603,11 @@ func (s *sketchInfo) addRow(
 	if allNulls {
 		s.numNulls++
 	}
-	s.sketch.Insert(*buf)
+	if s.sketchNew != nil {
+		s.sketchNew.Insert(*buf)
+	} else {
+		s.sketchOld.Insert(*buf)
+	}
 	return nil
 }
 
@@ -657,7 +678,11 @@ func (s *sketchInfo) addRowLegacy(
 		// be uniformly distributed in the 2^64 range). Experiments (on tpcc
 		// order_line) with simplistic functions yielded bad results.
 		binary.LittleEndian.PutUint64(*buf, uint64(val))
-		s.sketch.Insert(*buf)
+		if s.sketchNew != nil {
+			s.sketchNew.Insert(*buf)
+		} else {
+			s.sketchOld.Insert(*buf)
+		}
 		return nil
 	}
 	isNull := true
@@ -690,6 +715,10 @@ func (s *sketchInfo) addRowLegacy(
 	if isNull {
 		s.numNulls++
 	}
-	s.sketch.Insert(*buf)
+	if s.sketchNew != nil {
+		s.sketchNew.Insert(*buf)
+	} else {
+		s.sketchOld.Insert(*buf)
+	}
 	return nil
 }

